x/outline-cli/main.go

```diff
@@ -0,0 +1,390 @@
+	tun, err := netlink.LinkByName(OUTLINE_TUN_NAME)
+	if err != nil {
+		fmt.Printf("fatal error: %v\n", err)
+		return err
+	}
+
+	dst, err := netlink.ParseIPNet(OUTLINE_GW_SUBNET)
+	if err != nil {
+		fmt.Printf("fatal error: %v\n", err)
+		return err
+	}
+	r := netlink.Route{
+		LinkIndex: tun.Attrs().Index,
+		Table:     OUTLINE_ROUTING_TABLE,
+		Dst:       dst,
+		Src:       net.ParseIP(OUTLINE_TUN_IP),
+		Scope:     netlink.SCOPE_LINK,
+	}
+	fmt.Printf("\trouting only from %v to %v through nic %v...\n", r.Src, r.Dst, r.LinkIndex)
+	err = netlink.RouteAdd(&r)
+	if err != nil {
+		fmt.Printf("fatal error: %v\n", err)
+		return err
+	}
+
+	r = netlink.Route{
+		LinkIndex: tun.Attrs().Index,
+		Table:     OUTLINE_ROUTING_TABLE,
+		Gw:        net.ParseIP(OUTLINE_GW_IP),
+	}
+	fmt.Printf("\tdefault routing entry via gw %v through nic %v...\n", r.Gw, r.LinkIndex)
+	err = netlink.RouteAdd(&r)
+	if err != nil {
+		fmt.Printf("fatal error: %v\n", err)
+		return err
+	}
+
+	fmt.Println("routing table has been successfully configured")
+	return nil
+}
+
+func cleanUpRouting() error {
+	fmt.Println("cleaning up outline routing table...")
+	filter := netlink.Route{Table: OUTLINE_ROUTING_TABLE}
+	routes, err := netlink.RouteListFiltered(netlink.FAMILY_V4, &filter, netlink.RT_FILTER_TABLE)
+	if err != nil {
+		fmt.Printf("fatal error: %v\n", err)
+		return err
+	}
+	var lastErr error = nil
+	for _, route := range routes {
+		if err := netlink.RouteDel(&route); err != nil {
+			fmt.Printf("fatal error: %v\n", err)
+			lastErr = err
+		}
+	}
+	if lastErr == nil {
+		fmt.Println("routing table has been reset")
+	}
+	return lastErr
+}
+
+func showAllRules() error {
+	rules, err := netlink.RuleList(netlink.FAMILY_ALL)
+	if err != nil {
+		fmt.Printf("fatal error: %v\n", err)
+		return err
+	}
+	for _, r := range rules {
+		fmt.Printf("\t%v\n", r)
+	}
+	return nil
+}
+
+func setupIpRule(svrIp string) (*netlink.Rule, error) {
+	fmt.Println("adding ip rule for outline routing table...")
+	dst, err := netlink.ParseIPNet(svrIp)
+	if err != nil {
+		fmt.Printf("fatal error: %v\n", err)
+		return nil, err
+	}
+	rule := netlink.NewRule()
+	rule.Priority = OUTLINE_ROUTING_PRIORITY
+	rule.Family = netlink.FAMILY_V4
+	rule.Table = OUTLINE_ROUTING_TABLE
+	rule.Dst = dst
+	rule.Invert = true
+	fmt.Printf("+from all not to %v via table %v...\n", rule.Dst, rule.Table)
+	err = netlink.RuleAdd(rule)
+	if err != nil {
+		fmt.Printf("fatal error: %v\n", err)
+		return nil, err
+	}
+	fmt.Println("ip rule for outline routing table created")
+	return rule, nil
+}
+
+func cleanUpRule(rule *netlink.Rule) error {
+	fmt.Println("cleaning up ip rule of routing table...")
+	err := netlink.RuleDel(rule)
+	if err != nil {
+		fmt.Printf("fatal error: %v\n", err)
+		return err
+	}
+	fmt.Println("ip rule of routing table deleted")
+	return nil
+}
+
+func startTun2Socks(tun *water.Interface, ip, pass string, port int) (network.IPDevice, error) {
```

Comment:
`tun` is not used.

I'd like to see a function `createIPDevice(transportConfig) (network.IPDevice, error)` that is independent of routing.